File: python/src/kanbus/hooks.py
"""Lifecycle hook runtime and built-in providers."""

from __future__ import annotations

import json
import os
import shutil
import subprocess
import time
from dataclasses import dataclass
from datetime import datetime, timezone
from enum import Enum
from pathlib import Path
from typing import Any, Protocol

import click

from kanbus.config_loader import ConfigurationError, load_project_configuration
from kanbus.models import HookDefinition, HooksConfiguration, IssueData
from kanbus.project import ProjectMarkerError, get_configuration_path

HOOK_SCHEMA_VERSION = "kanbus.hooks.v1"


class HookPhase(str, Enum):
    """Lifecycle hook phase."""

    BEFORE = "before"
    AFTER = "after"


class HookEvent(str, Enum):
    """Normalized lifecycle hook events."""

    ISSUE_CREATE = "issue.create"
    ISSUE_UPDATE = "issue.update"
    ISSUE_CLOSE = "issue.close"
    ISSUE_DELETE = "issue.delete"
    ISSUE_COMMENT = "issue.comment"
    ISSUE_DEPENDENCY = "issue.dependency"
    ISSUE_PROMOTE = "issue.promote"
    ISSUE_LOCALIZE = "issue.localize"
    ISSUE_SHOW = "issue.show"
    ISSUE_LIST = "issue.list"
    ISSUE_READY = "issue.ready"


MUTATING_EVENTS = {
    HookEvent.ISSUE_CREATE,
    HookEvent.ISSUE_UPDATE,
    HookEvent.ISSUE_CLOSE,
    HookEvent.ISSUE_DELETE,
    HookEvent.ISSUE_COMMENT,
    HookEvent.ISSUE_DEPENDENCY,
    HookEvent.ISSUE_PROMOTE,
    HookEvent.ISSUE_LOCALIZE,
}

POLICY_OPERATION_BY_EVENT = {
    HookEvent.ISSUE_CREATE: "create",
    HookEvent.ISSUE_UPDATE: "update",
    HookEvent.ISSUE_CLOSE: "close",
    HookEvent.ISSUE_DELETE: "delete",
    HookEvent.ISSUE_SHOW: "view",
    HookEvent.ISSUE_LIST: "list",
    HookEvent.ISSUE_READY: "ready",
}


class HookExecutionError(RuntimeError):
    """Raised when a blocking hook fails."""


@dataclass(frozen=True)
class HookInvocation:
    """Canonical payload envelope delivered to hook handlers."""

    schema_version: str
    phase: HookPhase
    event: HookEvent
    timestamp: str
    actor: str
    mode: dict[str, Any]
    operation: dict[str, Any]

    def to_payload(self) -> dict[str, Any]:
        return {
            "schema_version": self.schema_version,
            "phase": self.phase.value,
            "event": self.event.value,
            "timestamp": self.timestamp,
            "actor": self.actor,
            "mode": self.mode,
            "operation": self.operation,
        }


@dataclass(frozen=True)
class HookResult:
    """Result metadata for one hook execution."""

    hook_id: str
    succeeded: bool
    timed_out: bool
    exit_code: int | None
    duration_ms: int
    message: str


class HookHandler(Protocol):
    """Hook handler execution contract."""

    def run(
        self,
        *,
        hook: HookDefinition,
        invocation: HookInvocation,
        project_root: Path,
        timeout_ms: int,
    ) -> HookResult: ...


class ExternalCommandHookHandler:
    """Execute external command hooks with JSON payload over stdin."""

    def run(
        self,
        *,
        hook: HookDefinition,
        invocation: HookInvocation,
        project_root: Path,
        timeout_ms: int,
    ) -> HookResult:
        env = os.environ.copy()
        env.update(hook.env)

        cwd = project_root
        if hook.cwd:
            candidate = Path(hook.cwd)
            cwd = candidate if candidate.is_absolute() else (project_root / candidate)

        payload = json.dumps(invocation.to_payload(), sort_keys=True, default=str)
        started = time.monotonic()
        try:
            completed = subprocess.run(
                hook.command,
                input=payload,
                text=True,
                capture_output=True,
                timeout=max(timeout_ms, 1) / 1000,
                cwd=cwd,
                env=env,
                check=False,
            )
        except subprocess.TimeoutExpired:
            duration_ms = int((time.monotonic() - started) * 1000)
            return HookResult(
                hook_id=hook.id,
                succeeded=False,
                timed_out=True,
                exit_code=None,
                duration_ms=duration_ms,
                message=f"timed out after {timeout_ms}ms",
            )
        except OSError as error:
            duration_ms = int((time.monotonic() - started) * 1000)
            return HookResult(
                hook_id=hook.id,
                succeeded=False,
                timed_out=False,
                exit_code=None,
                duration_ms=duration_ms,
                message=str(error),
            )

        duration_ms = int((time.monotonic() - started) * 1000)
        succeeded = completed.returncode == 0
        message = (
            "ok"
            if succeeded
            else (completed.stderr.strip() or f"exit code {completed.returncode}")
        )
        return HookResult(
            hook_id=hook.id,
            succeeded=succeeded,
            timed_out=False,
            exit_code=completed.returncode,
            duration_ms=duration_ms,
            message=message,
        )


def hooks_globally_disabled(no_hooks: bool = False) -> bool:
    """Return True if hooks are globally disabled by flag or env."""
    if no_hooks:
        return True
    raw = os.environ.get("KANBUS_NO_HOOKS", "").strip().lower()
    return raw in {"1", "true", "yes", "on"}


def serialize_issue(issue: IssueData | None) -> dict[str, Any] | None:
    """Serialize an issue snapshot for hook payloads."""
    if issue is None:
        return None
    return issue.model_dump(by_alias=True, mode="json")


def list_hooks(root: Path) -> list[dict[str, Any]]:
    """List configured hooks for CLI inspection."""
    hooks_config, _ = _resolve_hook_configuration(root)
    rows: list[dict[str, Any]] = []
    for phase, event_map in (
        (HookPhase.BEFORE, hooks_config.before),
        (HookPhase.AFTER, hooks_config.after),
    ):
        for event_name in sorted(event_map):
            hooks = event_map[event_name]
            for hook in hooks:
                event = _parse_hook_event(event_name)
                timeout_ms = hook.timeout_ms or hooks_config.default_timeout_ms
                rows.append(
                    {
                        "source": "external",
                        "phase": phase.value,
                        "event": event_name,
                        "id": hook.id,
                        "command": " ".join(hook.command),
                        "blocking": _effective_blocking(phase, event, hook),
                        "timeout_ms": timeout_ms,
                    }
                )

    for event in _policy_events():
        rows.append(
            {
                "source": "built-in",
                "phase": HookPhase.AFTER.value,
                "event": event.value,
                "id": "policy-guidance",
                "command": "<built-in>",
                "blocking": False,
                "timeout_ms": None,
            }
        )
    return rows


def validate_hooks(root: Path) -> list[str]:
    """Return validation problems for configured hooks."""
    hooks_config, project_root = _resolve_hook_configuration(root)
    issues: list[str] = []

    for phase_name, event_map in (
        (HookPhase.BEFORE.value, hooks_config.before),
        (HookPhase.AFTER.value, hooks_config.after),
    ):
        for event_name, hooks in event_map.items():
            event = _parse_hook_event(event_name)
            if event is None:
                issues.append(f"hooks.{phase_name}.{event_name}: unknown event")
                continue
            if not hooks:
                issues.append(f"hooks.{phase_name}.{event_name}: empty hook list")
                continue

            seen: set[str] = set()
            for hook in hooks:
                if hook.id in seen:
                    issues.append(
                        f"hooks.{phase_name}.{event_name}: duplicate hook id '{hook.id}'"
                    )
                seen.add(hook.id)

                executable = hook.command[0].strip() if hook.command else ""
                if not executable:
                    issues.append(
                        f"hooks.{phase_name}.{event_name}.{hook.id}: command is empty"
                    )
                    continue

                if _looks_like_path(executable):
                    candidate = Path(executable)
                    resolved = (
                        candidate
                        if candidate.is_absolute()
                        else (project_root / candidate)
                    )
                    if not resolved.exists():
                        issues.append(
                            f"hooks.{phase_name}.{event_name}.{hook.id}: command not found at {resolved}"
                        )
                elif shutil.which(executable) is None:
                    issues.append(
                        f"hooks.{phase_name}.{event_name}.{hook.id}: command '{executable}' is not on PATH"
                    )

                if hook.cwd:
                    cwd = Path(hook.cwd)
                    resolved_cwd = cwd if cwd.is_absolute() else (project_root / cwd)
                    if not resolved_cwd.exists() or not resolved_cwd.is_dir():
                        issues.append(
                            f"hooks.{phase_name}.{event_name}.{hook.id}: cwd '{resolved_cwd}' does not exist"
                        )
    return issues


def run_lifecycle_hooks(
    *,
    root: Path,
    phase: HookPhase,
    event: HookEvent,
    actor: str,
    beads_mode: bool,
    operation: dict[str, Any] | None = None,
    issues_for_policy: list[IssueData] | None = None,
    no_hooks: bool = False,
    no_guidance: bool = False,
) -> None:
    """Run external hooks and built-in policy provider for one lifecycle event."""
    if hooks_globally_disabled(no_hooks):
        return

    hooks_config, project_root = _resolve_hook_configuration(root)
    if not hooks_config.enabled:
        return
    if beads_mode and not hooks_config.run_in_beads_mode:
        return

    invocation = HookInvocation(
        schema_version=HOOK_SCHEMA_VERSION,
        phase=phase,
        event=event,
        timestamp=_now_utc_iso(),
        actor=actor,
        mode={
            "beads_mode": beads_mode,
            "project_root": str(project_root),
            "working_directory": str(Path.cwd()),
            "runtime": "python",
        },
        operation=operation or {},
    )

    handler = ExternalCommandHookHandler()
    hook_definitions = _hooks_for_event(hooks_config, phase, event)
    for hook in hook_definitions:
        timeout_ms = hook.timeout_ms or hooks_config.default_timeout_ms
        result = handler.run(
            hook=hook,
            invocation=invocation,
            project_root=project_root,
            timeout_ms=timeout_ms,
        )
        blocking = _effective_blocking(phase, event, hook)
        if result.succeeded:
            continue
        if phase == HookPhase.BEFORE and blocking:
            raise HookExecutionError(
                f"blocking hook '{hook.id}' failed for {event.value}: {result.message}"
            )
        click.echo(
            (
                f"Hook warning ({event.value}/{phase.value}/{hook.id}): "
                f"{result.message}"
            ),
            err=True,
        )

    if phase == HookPhase.AFTER:
        _run_policy_guidance_provider(
            project_root=project_root,
            event=event,
            issues_for_policy=issues_for_policy or [],
            beads_mode=beads_mode,
            no_guidance=no_guidance,
        )


def _resolve_hook_configuration(root: Path) -> tuple[HooksConfiguration, Path]:
    try:
        config_path = get_configuration_path(root)
        configuration = load_project_configuration(config_path)
        return configuration.hooks, config_path.parent
    except (ConfigurationError, ProjectMarkerError, OSError):
        return HooksConfiguration(), root


def _hooks_for_event(
    hooks_config: HooksConfiguration,
    phase: HookPhase,
    event: HookEvent,
) -> list[HookDefinition]:
    phase_map = hooks_config.before if phase == HookPhase.BEFORE else hooks_config.after
    return list(phase_map.get(event.value, []))


def _effective_blocking(
    phase: HookPhase,
    event: HookEvent | None,
    hook: HookDefinition,
) -> bool:
    if phase == HookPhase.AFTER:
        return False
    if hook.blocking is not None:
        return hook.blocking
    return event in MUTATING_EVENTS


def _policy_events() -> list[HookEvent]:
    return sorted(POLICY_OPERATION_BY_EVENT.keys(), key=lambda value: value.value)


def _run_policy_guidance_provider(
    *,
    project_root: Path,
    event: HookEvent,
    issues_for_policy: list[IssueData],
    beads_mode: bool,
    no_guidance: bool,
) -> None:
    if beads_mode or not issues_for_policy:
        return
    operation_name = POLICY_OPERATION_BY_EVENT.get(event)
    if operation_name is None:
        return
    try:
        from kanbus.policy_context import PolicyOperation
        from kanbus.policy_guidance import emit_guidance_for_issues

        emit_guidance_for_issues(
            project_root,
            issues_for_policy,
            PolicyOperation(operation_name),
            no_guidance=no_guidance,
        )
    except Exception as error:  # pragma: no cover - intentionally non-blocking
        click.echo(
            f"Hook warning ({event.value}/after/policy-guidance): {error}",
            err=True,
        )


def _looks_like_path(command: str) -> bool:
    return command.startswith(".") or "/" in command or "\\" in command


def _parse_hook_event(value: str) -> HookEvent | None:
    try:
        return HookEvent(value)
    except ValueError:
        return None


def _now_utc_iso() -> str:
    return datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace(
        "+00:00", "Z"
    )

File: rust/src/hooks.rs
//! Lifecycle hook runtime and built-in providers.

use std::collections::{HashMap, HashSet};
use std::io::{Read, Write};
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};
use std::time::{Duration, Instant};

use chrono::{SecondsFormat, Utc};
use serde_json::{json, Value};

use crate::config_loader::load_project_configuration;
use crate::error::KanbusError;
use crate::file_io::get_configuration_path;
use crate::models::{HookDefinition, HooksConfiguration, IssueData};
use crate::rich_text_signals::emit_stderr_line;
use crate::users::get_current_user;

const HOOK_SCHEMA_VERSION: &str = "kanbus.hooks.v1";

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum HookPhase {
    Before,
    After,
}

impl HookPhase {
    fn as_str(self) -> &'static str {
        match self {
            HookPhase::Before => "before",
            HookPhase::After => "after",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum HookEvent {
    IssueCreate,
    IssueUpdate,
    IssueClose,
    IssueDelete,
    IssueComment,
    IssueDependency,
    IssuePromote,
    IssueLocalize,
    IssueShow,
    IssueList,
    IssueReady,
}

impl HookEvent {
    pub fn as_str(self) -> &'static str {
        match self {
            HookEvent::IssueCreate => "issue.create",
            HookEvent::IssueUpdate => "issue.update",
            HookEvent::IssueClose => "issue.close",
            HookEvent::IssueDelete => "issue.delete",
            HookEvent::IssueComment => "issue.comment",
            HookEvent::IssueDependency => "issue.dependency",
            HookEvent::IssuePromote => "issue.promote",
            HookEvent::IssueLocalize => "issue.localize",
            HookEvent::IssueShow => "issue.show",
            HookEvent::IssueList => "issue.list",
            HookEvent::IssueReady => "issue.ready",
        }
    }

    fn parse(value: &str) -> Option<Self> {
        match value {
            "issue.create" => Some(HookEvent::IssueCreate),
            "issue.update" => Some(HookEvent::IssueUpdate),
            "issue.close" => Some(HookEvent::IssueClose),
            "issue.delete" => Some(HookEvent::IssueDelete),
            "issue.comment" => Some(HookEvent::IssueComment),
            "issue.dependency" => Some(HookEvent::IssueDependency),
            "issue.promote" => Some(HookEvent::IssuePromote),
            "issue.localize" => Some(HookEvent::IssueLocalize),
            "issue.show" => Some(HookEvent::IssueShow),
            "issue.list" => Some(HookEvent::IssueList),
            "issue.ready" => Some(HookEvent::IssueReady),
            _ => None,
        }
    }

    fn is_mutating(self) -> bool {
        matches!(
            self,
            HookEvent::IssueCreate
                | HookEvent::IssueUpdate
                | HookEvent::IssueClose
                | HookEvent::IssueDelete
                | HookEvent::IssueComment
                | HookEvent::IssueDependency
                | HookEvent::IssuePromote
                | HookEvent::IssueLocalize
        )
    }
}

#[derive(Debug, Clone, Copy)]
pub struct HookExecutionOptions {
    pub beads_mode: bool,
    pub no_hooks: bool,
    pub no_guidance: bool,
}

#[derive(Debug, Clone)]
struct HookResult {
    succeeded: bool,
    timed_out: bool,
    exit_code: Option<i32>,
    duration_ms: u64,
    message: String,
}

#[derive(Debug, Clone)]
pub struct HookListRow {
    pub source: String,
    pub phase: String,
    pub event: String,
    pub id: String,
    pub command: String,
    pub blocking: bool,
    pub timeout_ms: Option<u64>,
}

pub fn hooks_globally_disabled(no_hooks: bool) -> bool {
    if no_hooks {
        return true;
    }
    let raw = std::env::var("KANBUS_NO_HOOKS")
        .unwrap_or_default()
        .trim()
        .to_ascii_lowercase();
    matches!(raw.as_str(), "1" | "true" | "yes" | "on")
}

pub fn serialize_issue(issue: &IssueData) -> Value {
    serde_json::to_value(issue).unwrap_or(Value::Null)
}

pub fn list_hooks(root: &Path) -> Vec<HookListRow> {
    let (hooks_config, _) = resolve_hook_configuration(root);
    let mut rows = Vec::new();

    for (phase, map) in [
        (HookPhase::Before, &hooks_config.before),
        (HookPhase::After, &hooks_config.after),
    ] {
        for (event_name, hooks) in map {
            let event = HookEvent::parse(event_name.as_str());
            for hook in hooks {
                let timeout_ms = hook.timeout_ms.or(Some(hooks_config.default_timeout_ms));
                rows.push(HookListRow {
                    source: "external".to_string(),
                    phase: phase.as_str().to_string(),
                    event: event_name.clone(),
                    id: hook.id.clone(),
                    command: hook.command.join(" "),
                    blocking: effective_blocking(phase, event, hook),
                    timeout_ms,
                });
            }
        }
    }

    for event in policy_event_map().keys() {
        rows.push(HookListRow {
            source: "built-in".to_string(),
            phase: HookPhase::After.as_str().to_string(),
            event: event.as_str().to_string(),
            id: "policy-guidance".to_string(),
            command: "<built-in>".to_string(),
            blocking: false,
            timeout_ms: None,
        });
    }

    rows
}

pub fn validate_hooks(root: &Path) -> Vec<String> {
    let (hooks_config, project_root) = resolve_hook_configuration(root);
    let mut issues = Vec::new();

    for (phase_name, event_map) in [
        (HookPhase::Before.as_str(), &hooks_config.before),
        (HookPhase::After.as_str(), &hooks_config.after),
    ] {
        for (event_name, hooks) in event_map {
            let event = HookEvent::parse(event_name.as_str());
            if event.is_none() {
                issues.push(format!("hooks.{phase_name}.{event_name}: unknown event"));
                continue;
            }
            if hooks.is_empty() {
                issues.push(format!("hooks.{phase_name}.{event_name}: empty hook list"));
                continue;
            }

            let mut seen = HashSet::new();
            for hook in hooks {
                if !seen.insert(hook.id.clone()) {
                    issues.push(format!(
                        "hooks.{phase_name}.{event_name}: duplicate hook id '{}'",
                        hook.id
                    ));
                }

                let Some(executable) = hook.command.first() else {
                    issues.push(format!(
                        "hooks.{phase_name}.{event_name}.{}: command is empty",
                        hook.id
                    ));
                    continue;
                };

                if looks_like_path(executable) {
                    let candidate = Path::new(executable);
                    let resolved = if candidate.is_absolute() {
                        candidate.to_path_buf()
                    } else {
                        project_root.join(candidate)
                    };
                    if !resolved.exists() {
                        issues.push(format!(
                            "hooks.{phase_name}.{event_name}.{}: command not found at {}",
                            hook.id,
                            resolved.display()
                        ));
                    }
                } else if !command_on_path(executable) {
                    issues.push(format!(
                        "hooks.{phase_name}.{event_name}.{}: command '{}' is not on PATH",
                        hook.id, executable
                    ));
                }

                if let Some(cwd) = hook.cwd.as_ref() {
                    let cwd_path = Path::new(cwd);
                    let resolved = if cwd_path.is_absolute() {
                        cwd_path.to_path_buf()
                    } else {
                        project_root.join(cwd_path)
                    };
                    if !resolved.exists() || !resolved.is_dir() {
                        issues.push(format!(
                            "hooks.{phase_name}.{event_name}.{}: cwd '{}' does not exist",
                            hook.id,
                            resolved.display()
                        ));
                    }
                }
            }
        }
    }

    issues
}

pub fn run_lifecycle_hooks(
    root: &Path,
    phase: HookPhase,
    event: HookEvent,
    operation: Value,
    issues_for_policy: &[IssueData],
    options: HookExecutionOptions,
) -> Result<(), KanbusError> {
    if hooks_globally_disabled(options.no_hooks) {
        return Ok(());
    }

    let (hooks_config, project_root) = resolve_hook_configuration(root);
    if !hooks_config.enabled {
        return Ok(());
    }
    if options.beads_mode && !hooks_config.run_in_beads_mode {
        return Ok(());
    }

    let actor = get_current_user();
    let invocation = json!({
        "schema_version": HOOK_SCHEMA_VERSION,
        "phase": phase.as_str(),
        "event": event.as_str(),
        "timestamp": now_utc_iso(),
        "actor": actor,
        "mode": {
            "beads_mode": options.beads_mode,
            "project_root": project_root.display().to_string(),
            "working_directory": std::env::current_dir().unwrap_or_else(|_| PathBuf::from(".")).display().to_string(),
            "runtime": "rust",
        },
        "operation": operation,
    });

    let hooks = hooks_for_event(&hooks_config, phase, event);
    for hook in hooks {
        let timeout_ms = hook
            .timeout_ms
            .unwrap_or(hooks_config.default_timeout_ms)
            .max(1);
        let result = run_external_hook(&hook, &project_root, &invocation, timeout_ms);
        let blocking = effective_blocking(phase, Some(event), &hook);
        if result.succeeded {
            continue;
        }
        if phase == HookPhase::Before && blocking {
            return Err(KanbusError::IssueOperation(format!(
                "blocking hook '{}' failed for {}: {}",
                hook.id,
                event.as_str(),
                result.message
            )));
        }
        emit_stderr_line(&format!(
            "Hook warning ({}/{}/{}): {} (timeout={}, exit_code={:?}, duration_ms={})",
            event.as_str(),
            phase.as_str(),
            hook.id,
            result.message,
            result.timed_out,
            result.exit_code,
            result.duration_ms
        ));
    }

    if phase == HookPhase::After {
        run_policy_provider(
            &project_root,
            event,
            issues_for_policy,
            options.beads_mode,
            options.no_guidance,
        );
    }

    Ok(())
}

fn run_policy_provider(
    root: &Path,
    event: HookEvent,
    issues_for_policy: &[IssueData],
    beads_mode: bool,
    no_guidance: bool,
) {
    if beads_mode || issues_for_policy.is_empty() {
        return;
    }
    let policy_map = policy_event_map();
    let Some(operation) = policy_map.get(&event) else {
        return;
    };
    crate::policy_guidance::emit_guidance_for_issues(
        root,
        issues_for_policy,
        operation.clone(),
        no_guidance,
    );
}

fn policy_event_map() -> HashMap<HookEvent, crate::policy_context::PolicyOperation> {
    use crate::policy_context::PolicyOperation;
    HashMap::from([
        (HookEvent::IssueCreate, PolicyOperation::Create),
        (HookEvent::IssueUpdate, PolicyOperation::Update),
        (HookEvent::IssueClose, PolicyOperation::Close),
        (HookEvent::IssueDelete, PolicyOperation::Delete),
        (HookEvent::IssueShow, PolicyOperation::View),
        (HookEvent::IssueList, PolicyOperation::List),
        (HookEvent::IssueReady, PolicyOperation::Ready),
    ])
}

fn resolve_hook_configuration(root: &Path) -> (HooksConfiguration, PathBuf) {
    let config_path = match get_configuration_path(root) {
        Ok(path) => path,
        Err(_) => return (HooksConfiguration::default(), root.to_path_buf()),
    };
    let project_root = config_path
        .parent()
        .map(Path::to_path_buf)
        .unwrap_or_else(|| root.to_path_buf());

    match load_project_configuration(&config_path) {
        Ok(configuration) => (configuration.hooks, project_root),
        Err(_) => (HooksConfiguration::default(), project_root),
    }
}

fn hooks_for_event(
    hooks_config: &HooksConfiguration,
    phase: HookPhase,
    event: HookEvent,
) -> Vec<HookDefinition> {
    let map = if phase == HookPhase::Before {
        &hooks_config.before
    } else {
        &hooks_config.after
    };
    map.get(event.as_str()).cloned().unwrap_or_default()
}

fn effective_blocking(phase: HookPhase, event: Option<HookEvent>, hook: &HookDefinition) -> bool {
    if phase == HookPhase::After {
        return false;
    }
    if let Some(blocking) = hook.blocking {
        return blocking;
    }
    event.map(HookEvent::is_mutating).unwrap_or(false)
}

fn run_external_hook(
    hook: &HookDefinition,
    project_root: &Path,
    invocation: &Value,
    timeout_ms: u64,
) -> HookResult {
    if hook.command.is_empty() {
        return HookResult {
            succeeded: false,
            timed_out: false,
            exit_code: None,
            duration_ms: 0,
            message: "command is empty".to_string(),
        };
    }

    let mut command = Command::new(&hook.command[0]);
    if hook.command.len() > 1 {
        command.args(&hook.command[1..]);
    }

    let cwd = hook
        .cwd
        .as_ref()
        .map(PathBuf::from)
        .map(|cwd| {
            if cwd.is_absolute() {
                cwd
            } else {
                project_root.join(cwd)
            }
        })
        .unwrap_or_else(|| project_root.to_path_buf());

    command
        .stdin(Stdio::piped())
        .stdout(Stdio::null())
        .stderr(Stdio::piped())
        .current_dir(cwd)
        .envs(std::env::vars());

    for (key, value) in &hook.env {
        command.env(key, value);
    }

    let started = Instant::now();
    let mut child = match command.spawn() {
        Ok(child) => child,
        Err(error) => {
            return HookResult {
                succeeded: false,
                timed_out: false,
                exit_code: None,
                duration_ms: started.elapsed().as_millis() as u64,
                message: error.to_string(),
            }
        }
    };

    if let Some(mut stdin) = child.stdin.take() {
        let payload = serde_json::to_string(invocation).unwrap_or_else(|_| "{}".to_string());
        if let Err(error) = stdin.write_all(payload.as_bytes()) {
            return HookResult {
                succeeded: false,
                timed_out: false,
                exit_code: None,
                duration_ms: started.elapsed().as_millis() as u64,
                message: error.to_string(),
            };
        }
    }

    let timeout = Duration::from_millis(timeout_ms.max(1));
    let mut timed_out = false;
    let exit_code: Option<i32>;

    loop {
        match child.try_wait() {
            Ok(Some(status)) => {
                exit_code = status.code();
                break;
            }
            Ok(None) => {
                if started.elapsed() >= timeout {
                    timed_out = true;
                    let _ = child.kill();
                    let _ = child.wait();
                    exit_code = None;
                    break;
                }
                std::thread::sleep(Duration::from_millis(10));
            }
            Err(error) => {
                return HookResult {
                    succeeded: false,
                    timed_out: false,
                    exit_code: None,
                    duration_ms: started.elapsed().as_millis() as u64,
                    message: error.to_string(),
                }
            }
        }
    }

    let mut stderr = String::new();
    if let Some(mut stderr_pipe) = child.stderr.take() {
        let _ = stderr_pipe.read_to_string(&mut stderr);
    }

    let duration_ms = started.elapsed().as_millis() as u64;
    let succeeded = !timed_out && exit_code == Some(0);
    let message = if timed_out {
        format!("timed out after {timeout_ms}ms")
    } else if succeeded {
        "ok".to_string()
    } else if !stderr.trim().is_empty() {
        stderr.trim().to_string()
    } else {
        format!("exit code {:?}", exit_code)
    };

    HookResult {
        succeeded,
        timed_out,
        exit_code,
        duration_ms,
        message,
    }
}

fn looks_like_path(command: &str) -> bool {
    command.starts_with('.') || command.contains('/') || command.contains('\\')
}

fn command_on_path(command: &str) -> bool {
    let Some(path) = std::env::var_os("PATH") else {
        return false;
    };
    for entry in std::env::split_paths(&path) {
        let candidate = entry.join(command);
        if candidate.exists() {
            return true;
        }
    }
    false
}

fn now_utc_iso() -> String {
    Utc::now().to_rfc3339_opts(SecondsFormat::Millis, true)
}
